File: go/sqs/handlers.py
import boto3
import json
import os

from ddtrace import tracer

client = boto3.client('sqs')
queue_url = os.environ['SQS_QUEUE_URL']
runtime = os.environ['AWS_EXECUTION_ENV'].replace('AWS_Lambda_', '')

def producer(event, context):
    msg = json.dumps({
            'runtime': runtime,
            'trace_id': _current_trace_id(),
    })
    print(f'sending sqs message {msg} to {queue_url}')
    client.send_message(QueueUrl=queue_url, MessageBody=msg)
    return {'statusCode': 200, 'body': 'ok'}

def _current_trace_id():
    ctx = tracer.current_trace_context()
    assert ctx, 'no trace context found!'
    print(f'found trace context {ctx}')
    return str(ctx.trace_id)

File: apps/trace-propagation/node/handlers.js
const tracer = require('dd-trace').init();
const { SQSClient, SendMessageCommand } = require("@aws-sdk/client-sqs");
const { sendDistributionMetric } = require('datadog-lambda-js');

const client = new SQSClient();
const queueUrls = process.env.SQS_QUEUE_URLS.split(',');
const runtime = process.env.AWS_EXECUTION_ENV.replace('AWS_Lambda_', '');

exports.producer = async function(event, context) {
  msg = JSON.stringify({
    runtime: runtime,
    trace_id: currentTraceId(),
  })
  for (url of queueUrls) {
    console.log(`sending sqs message ${msg} to ${url}`);
    await client.send(new SendMessageCommand({
      MessageBody: msg,
      QueueUrl: url,
    }));
  }
  return {'statusCode': 200, 'body': 'ok'};
}

exports.consumer = async function(event, context) {
  const traceId = currentTraceId();
  event.Records.forEach(record => {
    const { body } = record;
    console.log(`received sqs message ${body}`);
    payload = JSON.parse(body);
    sendDistributionMetric(
      'trace_context.propagated.sqs', 1,
      `consumer_runtime:${runtime}`,
      `producer_runtime:${payload.runtime}`,
      traceId == payload.trace_id ? 'success:true' : 'success:false',
      'transport:sqs',
    );
  });
  return {'statusCode': 200, 'body': 'ok'};
}

function currentTraceId() {
  const ctx = tracer.scope().active()?.context();
  console.log(`found trace context: traceId=${ctx.toTraceId()} spanId=${ctx.toSpanId()}`);
  return ctx.toTraceId();
}
